nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/QueryDatabaseTable.java

```diff
@@ -0,0 +1,607 @@
+                                if (maxValueString != null) {
+                                    maxTimeValue = java.sql.Time.valueOf(maxValueString);
+                                }
+                                if (maxTimeValue == null || colTimeValue.after(maxTimeValue)) {
+                                    newColMap.put(colName, colTimeValue.toString());
+                                }
+                                break;
+
+                            case TIMESTAMP:
+                                if ("Oracle".equals(preProcessStrategy)) {
+                                    Date rawColOracleTimestampValue = resultSet.getDate(i);
+                                    java.sql.Date oracleTimestampValue = new java.sql.Date(rawColOracleTimestampValue.getTime());
+                                    java.sql.Date maxOracleTimestampValue = null;
+                                    if (maxValueString != null) {
+                                        maxOracleTimestampValue = java.sql.Date.valueOf(maxValueString);
+                                    }
+                                    if (maxOracleTimestampValue == null || oracleTimestampValue.after(maxOracleTimestampValue)) {
+                                        newColMap.put(colName, oracleTimestampValue.toString());
+                                    }
+                                } else {
+                                    Timestamp rawColTimestampValue = resultSet.getTimestamp(i);
+                                    java.sql.Timestamp colTimestampValue = new java.sql.Timestamp(rawColTimestampValue.getTime());
+                                    java.sql.Timestamp maxTimestampValue = null;
+                                    if (maxValueString != null) {
+                                        maxTimestampValue = java.sql.Timestamp.valueOf(maxValueString);
+                                    }
+                                    if (maxTimestampValue == null || colTimestampValue.after(maxTimestampValue)) {
+                                        newColMap.put(colName, colTimestampValue.toString());
+                                    }
+                                }
+                                break;
+
+                            case BIT:
+                            case BOOLEAN:
+                            case BINARY:
+                            case VARBINARY:
+                            case LONGVARBINARY:
+                            case ARRAY:
+                            case BLOB:
+                            case CLOB:
+                            default:
+                                throw new IOException("Type " + meta.getColumnTypeName(i) + " is not valid for maintaining maximum value");
+                        }
+                    }
+                }
+                // Set the new max values in the map
+                stateManager.setState(newColMap, Scope.CLUSTER);
```

Comment:
This is also making a call to ZooKeeper for each row in the ResultSet. I think we have to refactor this so that we can return just a single Map<String, String> and then store only the last one via the StateManager.